src/main/java/org/commonjava/util/partyline/impl/infinispan/model/InfinispanJFS.java

```diff
@@ -0,0 +1,213 @@
+
+    public InfinispanJFS( final String nodeKey, final Cache<String, FileMeta> metadataCache,
+                          final Cache<UUID, FileBlock> blockCache )
+    {
+        this.nodeKey = nodeKey;
+        this.metadataCache = metadataCache;
+        this.blockCache = blockCache;
+    }
+
+    @Override
+    public JoinableFile getFile( final File file, final LocalLockOwner lockOwner, final StreamCallbacks callbacks,
+                                 final boolean doOutput, final ReentrantOperationLock opLock ) throws IOException
+    {
+        return new InfinispanJF( file, lockOwner, callbacks, doOutput, opLock, this );
+    }
+
+    @Override
+    public LocalLockManager getLocalLockManager()
+    {
+        return null;
+    }
+
+    FileBlock getNextBlock( final FileBlock prevBlock, final FileMeta metadata) throws IOException
+    {
+        synchronized ( InfinispanJFS.this )
+        {
+            // TODO: make sure next will exist
+            UUID next = prevBlock.getNextBlockID();
+            FileBlock nextBlock = null;
+            while ( nextBlock == null )
+            {
+                nextBlock = blockCache.get( next );
+                if (nextBlock == null)
+                {
+                    try
+                    {
+                        lockManager.reentrantSynchronous( metadata.getFilePath(), ( opLock ) -> {
+                            // setup a cache listener for this UUID, and wait in a timed loop for it to return
+                            ClusterListener clusterListener = new ClusterListener( next, opLock );
+                            clusterListener.listenToCache( blockCache );
+                            opLock.await( WAIT_TIMEOUT );
+
+                            return null;
+                        } );
+                    }
+                    catch ( final InterruptedException e )
+                    {
+                        return null;
+                    }
+                }
+
+            }
+
+            return nextBlock;
+        }
+    }
+
+    void pushNextBlock( final FileBlock prevBlock, final FileBlock nextBlock, final FileMeta metadata )
+                    throws IOException
+    {
+        try
+        {
+            lockManager.reentrantSynchronous( metadata.getFilePath(), ( opLock ) -> {
+                blockCache.put( prevBlock.getBlockID(), prevBlock );
+                blockCache.put( nextBlock.getBlockID(), nextBlock );
+                metadataCache.put( metadata.getFilePath(), metadata );
+
+                return null;
+            } );
+        }
+        catch ( InterruptedException e )
+        {
+            throw new IOException( "Thread interrupted while adding next block to file: " + metadata.getFilePath(), e );
+        }
+    }
+
+    void updateBlock( final FileBlock block ) throws IOException
+    {
+        blockCache.put( block.getBlockID(), block );
+    }
+
+    void close( FileMeta metadata ) throws IOException
```

Comment:
We can pass in LocalLockOwner as a parameter here and avoid the need to store it in a FileMeta field. This will simplify the domain model we share (FileMeta) and avoid cases where LocalLockOwner is missing for a deserialized FileMeta. The InfinispanJF should contain a field to LocalLockOwner that we can use.